### PR TITLE
[mlir][Vector] Fix vector.extract lowering to llvm for 0-d vectors

### PR DESCRIPTION
**Pull request by Groverkss**
The current implementation of lowering to llvm for vector.extract incorrectly assumes that if the number of indices is zero, the operation can be folded away. This PR removes this condition and relies on the folder to do it instead.

This PR also unifies the logic for scalar extracts and slice extracts, which as a side effect also enables vector.extract lowering for n-d vector.extract with dynamic inner most dimension. (This was only prevented by a conservative check in the old implementation)

### CODE REVIEW

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1277]: [nit] it's the indices that are dynamic rather than dims

**Review by banach-space — COMMENTED** [review]
Just a couple of drive-by comments

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (kuhar) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1124]:
```suggestion
      if (!llvm::all_of(position, llvm::IsaPred<Attribute>)) {
```

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (kuhar) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1109]:
```suggestion
    // a scalar if the extract is full rank, i.e., the number of indices is equal
```

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (kuhar) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1277]:
```suggestion
// Multi-dim vectors are supported if the innermost dimension is dynamic.
```

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (kuhar) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1111]: Thanks for making these two modes very explicit, it makes it much easier to follow the logic.

**Review by kuhar — APPROVED** [review]
LGTM % nits. Please wait for an approval from someone who has stake in llvm lowering before merging this.

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (banach-space) [inline mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1107]: At the LLVM level, we don't truly "extract slices"—instead, we extract a member from an aggregate. For reference:

* [LLVM LangRef: Aggregate Operations](https://llvm.org/docs/LangRef.html#aggregate-operations)

According to the LangRef:

* `extractelement` extracts a scalar.
* `extractvalue` extracts a value, which in this case could be an aggregate or a vector.

This duality is particularly confusing because "Extract" exists at both the `Vector` and `LLVM` levels. To address this, perhaps we could rename the predicates for clarity:

* Rename `isSlicingExtract` -> `extractsAggregate` or `extractsValue`.
* Rename `isScalarExtract` -> `extractsScalar` or `extractsElement`.

Naming is hard 🤷🏻‍♂️. My suggestion would be to aim for terms that:

* Align closely with existing taxonomy/terminology.
* Clearly distinguish whether "extract" refers to `Vector` or `LLVM` (i.e., pre-lowering vs. post-lowering extraction).

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (Groverkss) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1099]: Right, I meant Vector -> LLVM lowering. I will clarify this. Also will change it to match the docs.

Comment (Groverkss) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1099]: Personally, I prefer "stacked 1-D" vectors as easier to understand though. But matching documentation is more important, so i will do that.

File: mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp
Comment (Groverkss) [reply mlir/lib/Conversion/VectorToLLVM/ConvertVectorToLLVM.cpp:1107]: Done. Let me know if it is clearer now :)

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1298]: [nit] "supported" is a bit ambiguous here. "Lowering supports extracting from multi-dim vectors with dynamic indices provided that _only_ the trailing index is dynamic." (I just struggle a bit with "innermost"  - "innermost" with respect to what? 😅 )

Feel free to re-use/modify/ignore.

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (banach-space) [inline mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1302]: [nit] Why not add types as well? These are significant, right?

**Review by banach-space — APPROVED** [review]
Thanks for the updates and sorry for missing this! LGTM, really appreciate the detailed documentation 🙏🏻 

I've left a few nits. Feel free to ignore.

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (Groverkss) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1298]: updated

File: mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir
Comment (Groverkss) [reply mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:1302]: I think the types being correct is tested elsewhere in the test file. These tests are just a indicator if it works or not, not if it is correct. I prefer restricting the scope of a particular test.

**Comment by bjacob** [comment]
Sent https://github.com/llvm/llvm-project/pull/128915 for review. Thanks again @Groverkss .

**Pull request merged**
